### ISSUE TITLE
Remove dependency to `@com_google_protobuf//src/google/protobuf/compiler:retention` without visibility

### ISSUE THREAD

**Issue by mering** [envoyproxy/envoy#37669]
*Description*:

Envoy API depends on the target `@com_google_protobuf//src/google/protobuf/compiler:retention` which does not grant visibility outside of Protobuf itself:
https://github.com/envoyproxy/envoy/blob/d1f6f3c669226f7d1dd99312b6e11b368879ca2b/api/bazel/cc_proto_descriptor_library/BUILD#L40

In the `WORKSPACE` model, this is patched out via https://github.com/envoyproxy/envoy/blob/d1f6f3c669226f7d1dd99312b6e11b368879ca2b/bazel/protobuf.patch#L116-L128

Downstream patches are not allowed with Bzlmod as this breaks automatic dependency resolution (see #34355 for a WIP PR adding Bzlmod support). We need to find a way to make envoy compatible with upstream Protobuf without patches or upstream necessary patches.

This is blocking addition of Bzlmod support. Bzlmod support was introduced as experimental in Bazel 5, fully introduced in Bazel 6, defaulted in Bazel 7, WORKSPACE support is disabled by default in Bazel 8 and will be removed in Bazel 9.

**Comment by phlax** [comment]
> Downstream patches are not allowed with Bzlmod 

@mering is this correct? my understanding is that upstream deps can be patched in much the same way as they can with workspace

**Comment by phlax** [comment]
> Now imagine both B and C patching D (differently). There is no way to resolve this problem automatically

for sure we have a lot of patches and regularly how to resolve their shortcomings

trying to understand the distinction of the root module ...

so eg if envoy deps on envoy_api, and envoy_api on protobuf - then any patches that envoy_api applies to protobuf would not happen automatically - envoy would need to explicitly declare the protobuf dep and add the patches that envoy_api has

is that not the same for workspace?

**Comment by phlax** [comment]
hmm - actually thinking about it - in workspace if we call `envoy_api_deps()` or somesuch that _would_ apply the patches

i think ive got it - so from our pov, as long as we are happy to carry whatever patches directly it wouldnt matter so much, but from _our_ downstreams pov it would make life more difficult (assuming they previously just called `envoy_deps()` or somesuch)

**Comment by mering** [comment]
> trying to understand the distinction of the root module ...
> 
> so eg if envoy deps on envoy_api, and envoy_api on protobuf - then any patches that envoy_api applies to protobuf would not happen automatically - envoy would need to explicitly declare the protobuf dep and add the patches that envoy_api has
> 
> is that not the same for workspace?

Yes this is correct. In WORKSPACE, every Bazel workspace (similar to root modules, basically where you run your Bazel commands) has to declare all dependencies including all transitive ones. As this adds up quite quickly, people started using macros to declare dependencies based on the definition from another repo. The problems of finding a suitable transitive dependency which fulfills the need of all consumers is usually a big mess, very time consuming and often requires adding even more patches on top.

So in your example, not only envoy as root module would need to add such patches but also every other dependency depending on enviy_api (like gRPC) would have to add these patches. Also this is a blocker of adding this dependency to BCR for easy consumption of users. So at least envoy_api should work without patches.

**Comment by phlax** [comment]
tbh - i think eliminating all of our existing patches _and_ any future need to ever require them is a losing battle

patches are generally regarded as a last resort but for so many things they are 100% necessary (without maintaining a fork or somesuch)

protobuf in particular is going to be more/less impossible to do that with

**Comment by mering** [comment]
For envoy_api the Protobuf visibility issue, which this issue is about, is the only patch (outside of this repo) required to make it build with Bzlmod.

**Comment by phlax** [comment]
not so familiar with BCR yet, so not clear how that works there - but i think, in general, we will need some way of documenting and making explicit any patches that we require

**Comment by mering** [comment]
You cannot upload to BCR if you require patches.

**Comment by mering** [comment]
So for this specific case, I see the following options:
1. Upstream the patch
2. Remove the need for this retention target
3. Fork the used functionality

**Comment by phlax** [comment]
yeah was just wondering similar - it seems its the  `StripSourceRetentionOptions` fun that is being used - perhaps upstream can be encouraged to move that to somewhere that is exported

**Comment by phlax** [comment]
altho in @keith 's PR he seems to have just removed the bazel dep -  not sure why that didnt fail

**Comment by mering** [comment]
> altho in @keith 's PR he seems to have just removed the bazel dep - not sure why that didnt fail

There just doesn't seem to be any users of that macro within `/api/`.
